pyacq/core/stream/stream.py

```diff
@@ -0,0 +1,294 @@
+        sample_rate: float or None
+            Sample rate of the stream in Hz.
+        sample_interval: float or None
+        sharedarray_shape: tuple
+            Shape of the SharedArray when using `transfermode = 'sharedarray'`.
+        ring_buffer_method: 'double' or 'single'
+            Method for the ring buffer when using `transfermode = 'sharedarray'`:
+            * 'single': a standard ring buffer.
+            * 'double': 2 ring buffers concatenated together. This ensures that
+            a continuous chunk exists in memory regardless of the sample position.
+            Note that, The ring buffer is along `timeaxis` for each case. And in case, of 'double', concatenated axis is also `timeaxis`.
+        shm_id : str or int (depending on platform)
+            id of the SharedArray when using `transfermode = 'sharedarray'`.
+
+        """
+        
+        self.params = dict(default_stream)
+        self.params.update(self.spec)
+        for k in kargs:
+            if k in self.spec:
+                assert kargs[k]==self.spec[k], \
+                    'Cannot configure {}={}; already in fixed in self.spec {}={}'.format(k, kargs[k], k, self.spec[k])
+        self.params.update(kargs)
+        
+        shape = self.params['shape']
+        assert shape[0] == -1 or shape[0] > 0, "First element in shape must be -1 or > 0."
+        for i in range(1, len(shape)):
+            assert shape[i] > 0, "Shape index %d must be > 0." % i
+        
+        if self.params['protocol'] in ('inproc', 'ipc'):
+            pipename = u'pyacq_pipe_'+''.join(random.SystemRandom().choice(string.ascii_uppercase + string.digits) for _ in range(24))
+            self.params['interface'] = pipename
+            self.url = '{protocol}://{interface}'.format(**self.params)
+        else:
+            self.url = '{protocol}://{interface}:{port}'.format(**self.params)
+        context = zmq.Context.instance()
+        self.socket = context.socket(zmq.PUB)
+        self.socket.bind(self.url)
+        self.addr = self.socket.getsockopt(zmq.LAST_ENDPOINT).decode()
+        self.port = self.addr.rpartition(':')[2]
+        self.params['port'] = self.port
+        
+        transfermode = self.params['transfermode']
+        if transfermode not in all_transfermodes:
+            raise ValueError("Unsupported transfer mode '%s'" % transfermode)
+        sender_class = all_transfermodes[transfermode][0]
+        self.sender = sender_class(self.socket, self.params)
+
+        self.configured = True
+        if self.node and self.node():
+            self.node().after_output_configure(self.name)
+
+    def send(self, data, index=None, **kargs):
+        """Send a data chunk and its frame index.
+        
+        Parameters
+        ----------
+        index: int
+            The absolute sample index. If the chunk contains multiple samples,
+            then this is the index of the last sample.
```

Comment:
Before this PR, index was the index of last sample+1, in short the upper bound excluded.
I think I prefer before...
Sorry to be so late for commenting that.

This is more intuitive for a python coder. The index was NOT (like here) the included upper bound but index is absolute length (so the excluded upper bound).

Do you known that we could banished from the python community with such a choice ?

What do you think.